Docs: Debugging Workflow (#1967)
ECP-WarpX/WarpX: Docs/source/usage/workflows/debugging.rst

.. _debugging_warpx:

Debugging the code
==================

Sometimes, the code does not give you the result that you are expecting.
This can be due to a variety of reasons, from misunderstandings or changes in the :ref:`input parameters <running-cpp-parameters>`, system specific quirks, or bugs.
You might also want to debug your code as you implement new features in WarpX during development.

This section gives a step-by-step guidance on how to systematically check what might be going wrong.

Debugging Workflow
------------------

> Try the following steps to debug a simulation.

Reviewer: ```suggestion
Try the following steps to debug a simulation:
```

#. Check the output text file, usually called ``output.txt``: are there warnings or errors present?
#. On an HPC system, look for the job output and error files, usually called ``WarpX.e...`` and ``WarpX.o...``.
   Read long messages from the top and follow potential guidance.
#. If your simulation already created output data files:

>    Check they look reasonable before the problem occurred; are the initial conditions of the simulation as you expected?

Reviewer: ```suggestion
   Check if they look reasonable before the problem occurred; are the initial conditions of the simulation as you expected?
```

   Do you spot numerical artifacts or instabilities that could point to missing resolution or unexpected/incompatible numerical parameters?
#. Did the job output files indicate a crash? Check the ``Backtrace.<mpirank>`` files for the location of the code that triggered the crash.
   Backtraces are read from bottom (high-level) to top (most specific line that crashed).

> #. In case of a crash, Backtraces can be more detailed if you :ref:`re-compile <install-developers>` with debug flags: ``-DCMAKE_BUILD_TYPE=Debug`` (this will make the simulation slower) and rerun.

Reviewer: ```suggestion
#. In case of a crash, Backtraces can be more detailed if you :ref:`re-compile <install-developers>` with debug flags: for example, try compiling with ``-DCMAKE_BUILD_TYPE=Debug`` (this will make the simulation slower) and rerun.
```

#. If debug builds are too costly, try instead compiling with ``-DAMReX_ASSERTIONS=ON`` to activate more checks and rerun.

> #. If the problem looks like an memory violating, this could be from an invalid field or particle index access.

Reviewer: ```suggestion
#. If the problem looks like a memory violation, this could be from an invalid field or particle index access.
```

>    Try compiling with ``-DAMReX_BOUND_CHECK=ON`` (this will make the simulation very slow), and re-run.

Reviewer: Just saw that you used "rerun" (without hyphen) in all other places, so I'd use it here as well, just for consistency:
```suggestion
   Try compiling with ``-DAMReX_BOUND_CHECK=ON`` (this will make the simulation very slow), and rerun.
```

> #. If the problem looks like a random memory might be used, try initializing memory with signaling Not-a-Number (NaN) values through the runtime option ``fab.init_snan = 1``.

Reviewer: Shall we mention also some of the other runtime options, like `amrex.fpe_trap_invalid`, `amrex.fpe_trap_zero`, or `amrex.fpe_trap_overflow`? Or is this what you refer to below when you talk about "AMReX-level signal handling"?

Author: I thought so too, but then the AMReX link below also explains then immediately. Still does not hurt to repeat?

Author: ```suggestion
#. If the problem looks like a random memory might be used, try initializing memory with signaling Not-a-Number (NaN) values through the runtime option ``fab.init_snan = 1``.
Further useful runtime options are ``amrex.fpe_trap_invalid``, ``amrex.fpe_trap_zero`` and ``amrex.fpe_trap_overflow`` (see details in the AMReX link below).
```

> #. On Nvidia GPUs, if you suspect the problem might be a race condition, set the environment variable ``export CUDA_LAUNCH_BLOCKING=1`` and rerun.

Reviewer 2: ```suggestion
#. On Nvidia GPUs, if you suspect the problem might be a race condition due to a missing host / device synchronization, set the environment variable ``export CUDA_LAUNCH_BLOCKING=1`` and rerun.
```

Reviewer 2: in other works it won't help with all race conditions, just ones arising from the asynchronous nature of CPU versus GPU execution.

Author: That's more precise, thanks!

> #. Consider simplifying your input options and readding more options after having found a working baseline.

Reviewer: ```suggestion
#. Consider simplifying your input options and reading more options after having found a working baseline.
```

Author: Oh, I meant "remove" and then "re-add" :D
```suggestion
#. Consider simplifying your input options and re-adding more options after having found a working baseline.
```
